### ISSUE TITLE
Performance metrics `listAccountHistory`

### ISSUE THREAD

**Issue by gamsterx** [DeFiCh/jellyfish#387]
Hi @fuxingloh  I have a question regarding the performance behavior of the `listAccountHistory`  command.

Node version: `1.7.10`
Yellyfish version: `0.12.0`

I´m using jellyfish to get my transactions und experimented with chopping up the requests in multiple batches (like pagination). I got some unexpected performance results and would like your opinion on it.

I executed the listaccounthistory with different  parameters.

Results:
The smaller the batch size and the more requests the longer it takes.

1. `listaccounthistory` with  `mine` `depth = blockCount`, and `limit = MAX_SAFE_INTEGER`

```shell
listAccountHistory from 915977 with depth 915977: 14022.044921875 ms
```

2. `listaccounthistory` with  `mine` `depth = 100000`, and `limit = MAX_SAFE_INTEGER`

```shell
listAccountHistory from 915934 with depth 100000: 13730.339111328125 ms
listAccountHistory from 815934 with depth 100000: 4062.484130859375 ms
listAccountHistory from 715934 with depth 100000: 4947.461181640625 ms
listAccountHistory from 615934 with depth 100000: 6408.5869140625 ms
listAccountHistory from 515934 with depth 100000: 7882.18994140625 ms
listAccountHistory from 415934 with depth 100000: 7924.23388671875 ms
listAccountHistory from 315934 with depth 100000: 8189.227294921875 ms
listAccountHistory from 215934 with depth 100000: 8182.885009765625 ms
listAccountHistory from 115934 with depth 100000: 7863.46484375 ms
listAccountHistory from 15934 with depth 100000: 8583.657958984375 ms
```

3. `listaccounthistory` with  `mine` `depth = 10000`, and `limit = MAX_SAFE_INTEGER`

```shell
listAccountHistory from 915945 with depth 10000: 2666.263916015625 ms
listAccountHistory from 905945 with depth 10000: 1992.656005859375 ms
listAccountHistory from 895945 with depth 10000: 2247.8408203125 ms
listAccountHistory from 885945 with depth 10000: 2371.30224609375 ms
listAccountHistory from 875945 with depth 10000: 2363.537109375 ms
listAccountHistory from 865945 with depth 10000: 4181.380126953125 ms
listAccountHistory from 855945 with depth 10000: 4110.50390625 ms
listAccountHistory from 845945 with depth 10000: 2984.93505859375 ms
listAccountHistory from 835945 with depth 10000: 3420.834228515625 ms
listAccountHistory from 825945 with depth 10000: 3645.53515625 ms
listAccountHistory from 815945 with depth 10000: 3359.8271484375 ms
listAccountHistory from 805945 with depth 10000: 3729.358154296875 ms
listAccountHistory from 795945 with depth 10000: 3862.964111328125 ms
listAccountHistory from 785945 with depth 10000: 4225.881103515625 ms
listAccountHistory from 775945 with depth 10000: 4061.76708984375 ms
listAccountHistory from 765945 with depth 10000: 5611.7939453125 ms
listAccountHistory from 755945 with depth 10000: 4824.673095703125 ms
listAccountHistory from 745945 with depth 10000: 4753.401123046875 ms
listAccountHistory from 735945 with depth 10000: 4553.934814453125 ms
listAccountHistory from 725945 with depth 10000: 4675.406982421875 ms
listAccountHistory from 715945 with depth 10000: 4825.72900390625 ms
listAccountHistory from 705945 with depth 10000: 5101.509765625 ms
listAccountHistory from 695945 with depth 10000: 5811.173095703125 ms
listAccountHistory from 685945 with depth 10000: 5372.3779296875 ms
listAccountHistory from 675945 with depth 10000: 5552.522216796875 ms
listAccountHistory from 665945 with depth 10000: 5545.52001953125 ms
listAccountHistory from 655945 with depth 10000: 5697.468017578125 ms
listAccountHistory from 645945 with depth 10000: 6289.509033203125 ms
listAccountHistory from 635945 with depth 10000: 6338.385986328125 ms
listAccountHistory from 625945 with depth 10000: 6167.596923828125 ms
listAccountHistory from 615945 with depth 10000: 6429.10888671875 ms
listAccountHistory from 605945 with depth 10000: 6491.8701171875 ms
listAccountHistory from 595945 with depth 10000: 7152.47900390625 ms
listAccountHistory from 585945 with depth 10000: 7183.89599609375 ms
listAccountHistory from 575945 with depth 10000: 6958.467041015625 ms
listAccountHistory from 565945 with depth 10000: 7039.7099609375 ms
listAccountHistory from 555945 with depth 10000: 7483.881103515625 ms
listAccountHistory from 545945 with depth 10000: 7580.385986328125 ms
listAccountHistory from 535945 with depth 10000: 7420.279052734375 ms
listAccountHistory from 525945 with depth 10000: 7609.2041015625 ms
listAccountHistory from 515945 with depth 10000: 8134.376953125 ms
listAccountHistory from 505945 with depth 10000: 8331.550048828125 ms
listAccountHistory from 495945 with depth 10000: 7985.054931640625 ms
listAccountHistory from 485945 with depth 10000: 8049.72802734375 ms
listAccountHistory from 475945 with depth 10000: 8192.090087890625 ms
listAccountHistory from 465945 with depth 10000: 8170.890869140625 ms
listAccountHistory from 455945 with depth 10000: 7948.885986328125 ms
listAccountHistory from 445945 with depth 10000: 7848.35009765625 ms
listAccountHistory from 435945 with depth 10000: 8146.239990234375 ms
listAccountHistory from 425945 with depth 10000: 8286.370849609375 ms
listAccountHistory from 415945 with depth 10000: 8125.609130859375 ms
listAccountHistory from 405945 with depth 10000: 8661.31103515625 ms
listAccountHistory from 395945 with depth 10000: 7890.9638671875 ms
listAccountHistory from 385945 with depth 10000: 8092.35302734375 ms
listAccountHistory from 375945 with depth 10000: 8639.26318359375 ms
listAccountHistory from 365945 with depth 10000: 8697.256103515625 ms
listAccountHistory from 355945 with depth 10000: 8252.044921875 ms
listAccountHistory from 345945 with depth 10000: 8064.43408203125 ms
listAccountHistory from 335945 with depth 10000: 8177.599853515625 ms
listAccountHistory from 325945 with depth 10000: 8151.36181640625 ms
listAccountHistory from 315945 with depth 10000: 8223.304931640625 ms
listAccountHistory from 305945 with depth 10000: 8187.89697265625 ms
listAccountHistory from 295945 with depth 10000: 8760.250244140625 ms
listAccountHistory from 285945 with depth 10000: 8960.788818359375 ms
listAccountHistory from 275945 with depth 10000: 8326.14697265625 ms
listAccountHistory from 265945 with depth 10000: 8114.1259765625 ms
listAccountHistory from 255945 with depth 10000: 8644.2890625 ms
listAccountHistory from 245945 with depth 10000: 8433.56689453125 ms
listAccountHistory from 235945 with depth 10000: 8042.621826171875 ms
listAccountHistory from 225945 with depth 10000: 8914.689697265625 ms
listAccountHistory from 215945 with depth 10000: 8413.66015625 ms
listAccountHistory from 205945 with depth 10000: 9076.339111328125 ms
listAccountHistory from 195945 with depth 10000: 8382.399169921875 ms
listAccountHistory from 185945 with depth 10000: 8451.809814453125 ms
listAccountHistory from 175945 with depth 10000: 9118.77783203125 ms
listAccountHistory from 165945 with depth 10000: 8758.65625 ms
listAccountHistory from 155945 with depth 10000: 9449.3720703125 ms
listAccountHistory from 145945 with depth 10000: 8990.319091796875 ms
listAccountHistory from 135945 with depth 10000: 8670.402099609375 ms
listAccountHistory from 125945 with depth 10000: 8720.38818359375 ms
listAccountHistory from 115945 with depth 10000: 8837.97412109375 ms
listAccountHistory from 105945 with depth 10000: 8860.684814453125 ms
listAccountHistory from 95945 with depth 10000: 8587.530029296875 ms
listAccountHistory from 85945 with depth 10000: 10138.6572265625 ms
listAccountHistory from 75945 with depth 10000: 8891.4580078125 ms
listAccountHistory from 65945 with depth 10000: 8956.875732421875 ms
listAccountHistory from 55945 with depth 10000: 9357.477783203125 ms
listAccountHistory from 45945 with depth 10000: 9548.032958984375 ms
listAccountHistory from 35945 with depth 10000: 8830.254150390625 ms
listAccountHistory from 25945 with depth 10000: 8745.31298828125 ms
listAccountHistory from 15945 with depth 10000: 8665.583251953125 ms
listAccountHistory from 5945 with depth 10000: 8244.087158203125 ms
```

**Comment by fuxingloh** [comment]
Hey @gamsterx 👋

In an effort to move fast in the early days we included product-facing indexes within the blockchain. These indexes were created to allow you to query your history ('mine') but in reality, the blockchain consensus logic doesn't need it and it's a waste of storage and complexity. The blockchain consensus has no concept of mine but rather script.

In order not to break compatibility, we remove those indexes but kept feature compatibility by iterating through a large dataset. Ultimately slated for deprecation as we don't need it moving forward.

If you want performance for that RPC, you have to index it yourself (cache it or create that index). Eventually, that will also be implemented in our new product-facing index. The project is called https://github.com/DeFiCh/whale in short it's the super index for DeFi Blockchain to simplify DeFi light implementation.

For your particular use case it's not yet implemented as we are have only implemented what we need for the light wallet development (probably sick of me saying this by now 😅). High-performance indexes that are currently are present here https://github.com/DeFiCh/whale/tree/main/packages/whale-api-client/src/api you can also fork the project and change the indexes to what you need, it's database agnostic. These endpoints will be also deployed as part of the https://github.com/DeFiCh/ocean project undertaking.

In summary, this is the project overview. I included other projects that are part of our new ecosystem undertaking that might interest you.

- `DeFiCh/ain` → Indexes data for consensus only, not good for product features.
- `DeFiCh/jellyfish` → RPC interfacing, wallet system, transaction crafting, transaction building, low-level blockchain, crypto logic, and other ecosystem tools.
- `DeFiCh/whale` → The super index for DeFi Blockchain to simplify DeFi light implementation. (product facing indexes)
- `DeFiCh/ocean` → Globally distributed and decentralized DeFiChain node infrastructure.
- `DeFiCh/playground` → DeFi playground for DeFi Blockchain engineers to build DeFi applications without running your own node.

**Comment by gamsterx** [comment]
Hey @fuxingloh

thank you for your detailed answer. I will take a look at the provided links and maybe come back with additional questions 😉 

I will close the issue.
